compiler/damlc/daml-stdlib-src/DA/Internal/Template/Functions.daml

```diff
@@ -196,16 +196,25 @@ class HasFromAnyContractKey t k | t -> k where
   -- | HIDE
   _fromAnyContractKey : proxy t -> Any -> Optional k
 
+-- | Constraint satisfied by choices from templates with key.
+type ChoiceByKey t k c r =
+  ( TemplateKey t k
+  , HasExerciseByKey t k c r
+  )
+
+-- | Exposes `exerciseByKey` function. Part of the `ChoiceByKey` constraint.
+class HasExerciseByKey t k c r | t -> k, t c -> r where
+   -- | HIDE
+   _exerciseByKey : proxy t -> k -> c -> Update r
+
 -- | Exercise a choice on the contract associated with the given key.
 --
 -- You must pass the `t` using an explicit type application. For
 -- instance, if you want to exercise a choice `Withdraw` on a contract of
 -- template `Account` given by its key `k`, you must call
 -- `exerciseByKey @Account k Withdraw`.
-exerciseByKey : forall t k c r. (HasFetchByKey t k, HasExercise t c r) => k -> c -> Update r
-exerciseByKey k c = do
-    (cid, _) <- fetchByKey @t k
-    exercise cid c
+exerciseByKey : forall t k c r. HasExerciseByKey t k c r => k -> c -> Update r
```

Comment:
Is it worth adding a test somewhere at the daml-lf level that a transaction actually contains an exercise by key node?

Reply:
683586700f1bbaf6f3f1222d9ac992c8b5a635b0